lib/tree.rb

```diff
@@ -16,42 +16,144 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key,value)
+
+    if @root.nil?
+      @root = new_node
+    else
+      add_node(@root, new_node)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_node(current, new_node)
+    return new_node if current.nil?
+
+    if new_node.key <= current.key
+      current.left = add_node(current.left, new_node)
+    else
+      current.right = add_node(current.right, new_node)
+    end
+
+    return current
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    current = @root
+    return nil if current.nil?
+    
+    until current == nil
+      if current.key == key
+        return current.value
+      elsif key < current.key
+        current = current.left
+      else
+        current = current.right
+      end
+    end
+    return nil
   end
 
+  #used the following as inspiration for the traversals: http://rubyalgorithms.com/binary_search_tree.html
+
   # Time Complexity: 
   # Space Complexity: 
   def inorder
```

Comment:
👍 Nice work with the block argument and the helper methods.   The space & time complexities would be O(n)